### DISCUSSION TITLE
PMS5003 giving out garbage values

### DISCUSSION THREAD

**Discussion by kengineer23** [micropython/micropython#13650] [ESP32]
I am using ESP32 WROOM DEVKIT to collect data from PMS5003 sensor. The sensor does send around 9-10 valid data before sending random garbage values. can y'all please review the code and suggest any improvements or it could just be a sensor issue??

----------
from machine import Pin, UART
import time

# Configure UART for PMS5003 sensor
uart = UART(2, 9600, tx=Pin(17), rx=Pin(16))

def read_pm25():
    while True:
       time.sleep(3)
       data = uart.read(32) ##read 32 bytes of data
       if data is None:
           print("False")
       else:
           print("True")
       if len(data) == 32 and data[0]==0x42 and data[1]== 0x4d:
           print("Collecting data")
           data_new=data[12] << 8 | data[13]
           return data_new
       else:
           print("Garbage data collected")
       #if data[0] == 0x42 and data[1] == 0x4d:
   # Calculate PM2.5 concentration
        #   data_new=data[6] << 8 | data[7]
         #  return data_new

time.sleep(30)
while True:
  pm25 = read_pm25()
  print(pm25, "ug/m3")

**Comment by robert-hh** [comment]
Pleas use lines with three backticks to enclose your code. The easiest way to get that is to highlight all code lines and push Ctrl-E.

**Comment by glenn20** [comment]
I would start by printing out the value of `data` when you get "garbage" to get clues on what is going wrong (eg. the device occasionally sends more than 32 characters). I'd also try setting the 3 second sleep much shorter in case you are getting some sort of buffer overrun. If you are using the PMS5003's active mode, the data from the sensor will be coming more quickly than this.

Also, when you get garbage, try checking `uart.any()` and flushing the input serial buffers to see if that restores good data. 

And, definitely do as @robert-hh suggests. It is not possible to be sure I understand your code correctly. Click on the `Welcome (Please read before posting)` link at the top of the [discussions page](https://github.com/orgs/micropython/discussions) for more information.
